frappe/email/receive.py

```diff
@@ -149,37 +153,86 @@ def get_messages(self):
 			else:
 				self.pop.quit()
 
-		return self.latest_messages
+		out = { "latest_messages": self.latest_messages }
+		if self.settings.use_imap:
+			out.update({
+				"uid_list": uid_list,
+				"seen_status": self.seen_status,
+				"unique_id_list": self.unique_id_list
+			})
+
+		return out
 
 	def get_new_mails(self):
 		"""Return list of new mails"""
 		if cint(self.settings.use_imap):
-			self.imap.select("Inbox")
-			response, message = self.imap.uid('search', None, "UNSEEN")
+			self.check_imap_uidvalidity()
+
+			self.imap.select("Inbox", readonly=True)
+			response, message = self.imap.uid('search', None, self.settings.email_sync_rule)
 			email_list =  message[0].split()
 		else:
 			email_list = self.pop.list()[1]
 
 		return email_list
 
+	def check_imap_uidvalidity(self):
+		# compare the UIDVALIDITY of email account and imap server
+		uid_validity = self.settings.uid_validity
+
+		responce, message = self.imap.status("Inbox", "(UIDVALIDITY UIDNEXT)")
+		current_uid_validity = self.parse_imap_responce("UIDVALIDITY", message[0])
+		if not current_uid_validity:
+			frappe.throw(_("Can not find UIDVALIDITY in imap status response"))
+
+		uidnext = int(self.parse_imap_responce("UIDNEXT", message[0]) or "1")
+		frappe.db.set_value("Email Account", self.settings.email_account, "uidnext", uidnext)
+
+		if not uid_validity or uid_validity != current_uid_validity:
```

Comment:
if not uid_validity should be the initial sync and behave separatly from a reindex should take the initial sync count and minus from uidnext